cmd/kube-dns/SPEC.md

```diff
@@ -0,0 +1,140 @@
+    - Answer Format:
+      - `_<port-x>._<proto-x>.<service>.<ns>.svc.<zone>. <ttl> IN SRV <weight> <priority> <port-number> <service>.<ns>.svc.<zone>.`
+    - Answer Example:
+```
+      _https._tcp.my-service.default.svc.cluster.local.  4 IN SRV 10 100 443 my-service.default.svc.cluster.local.
+      _http._tcp.my-service.default.svc.cluster.local.   4 IN SRV 10 100 80  my-service.default.svc.cluster.local.
+      _syslog._udp.my-service.default.svc.cluster.local. 4 IN SRV 10 100 514 my-service.default.svc.cluster.local.
+```
+  - If a Service named `<service>` exists in `<ns>` without a ClusterIP and with _ready_ endpoints, the result in the Answer section will be one SRV record for each _ready_ endpoint and port combination. Thus, if there are **N** ports for the service and **M** _ready_ endpoints, there will be **N** :heavy_multiplication_x: **M** records in the Answer section. The name of each record will be `_<port>._<proto>.<service>.<ns>.svc.<zone>`. Here, `<port>` is the name of the port if it has one, and empty if it does not. `<proto>` is either “udp” or “tcp” depending on the port protocol. There will therefore be **M** records with the same name, one for each _ready_ endpoint. The records will contain the port number of the `<port>` and the name of an A record `<hostname>.<service>.<ns>.svc.<zone>` where `<hostname>` is the endpoint _hostname_. This name will resolve to an A record with the IP address of the specific _ready_ endpoint (which may be included in the Additional section). The priority and weight are not prescribed by this specification.
+    - Answer Format:
+      - `_<port-x>._<proto-x>.<service>.<ns>.svc.<zone>. <ttl> IN SRV <weight> <priority> <port-number> <service>.<ns>.svc.<zone>.`
+    - Answer Example:
+```
+      _https._tcp.headless.default.svc.cluster.local.  4 IN SRV 10 100 443 my-pet-1.headless.default.svc.cluster.local.
+      _http._tcp.headless.default.svc.cluster.local.   4 IN SRV 10 100 80  my-pet-1.headless.default.svc.cluster.local.
+      _syslog._udp.headless.default.svc.cluster.local. 4 IN SRV 10 100 514 my-pet-1.headless.default.svc.cluster.local.
+      _https._tcp.headless.default.svc.cluster.local.  4 IN SRV 10 100 443 my-pet-2.headless.default.svc.cluster.local.
+      _http._tcp.headless.default.svc.cluster.local.   4 IN SRV 10 100 80  my-pet-2.headless.default.svc.cluster.local.
+      _syslog._udp.headless.default.svc.cluster.local. 4 IN SRV 10 100 514 my-pet-2.headless.default.svc.cluster.local.
+      _https._tcp.headless.default.svc.cluster.local.  4 IN SRV 10 100 443 438934893.headless.default.svc.cluster.local.
+      _http._tcp.headless.default.svc.cluster.local.   4 IN SRV 10 100 80  438934893.headless.default.svc.cluster.local.
+      _syslog._udp.headless.default.svc.cluster.local. 4 IN SRV 10 100 514 438934893.headless.default.svc.cluster.local.
+```
```

Comment:
kube-dns doesn't do this. The LHS here instead would all just be like the question `headless.default.svc.cluster.local.` which means you don't know which port/IP combo goes with each service and protocol. It also puts on the arbitrary identifier.

```
/ # dig headless.default.svc.cluster.local. srv

; <<>> DiG 9.10.4-P3 <<>> headless.default.svc.cluster.local. srv
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 27032
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 5, AUTHORITY: 0, ADDITIONAL: 5

;; QUESTION SECTION:
;headless.default.svc.cluster.local. IN	SRV

;; ANSWER SECTION:
headless.default.svc.cluster.local. 30 IN SRV	10 20 0 3862303234343538.headless.default.svc.cluster.local.
headless.default.svc.cluster.local. 30 IN SRV	10 20 0 6533653966373130.headless.default.svc.cluster.local.
headless.default.svc.cluster.local. 30 IN SRV	10 20 0 6463313736393962.headless.default.svc.cluster.local.
headless.default.svc.cluster.local. 30 IN SRV	10 20 0 6638383363643661.headless.default.svc.cluster.local.
headless.default.svc.cluster.local. 30 IN SRV	10 20 0 6337356433333633.headless.default.svc.cluster.local.

;; ADDITIONAL SECTION:
3862303234343538.headless.default.svc.cluster.local. 30	IN A 10.2.10.72
6533653966373130.headless.default.svc.cluster.local. 30	IN A 10.2.6.51
6463313736393962.headless.default.svc.cluster.local. 30	IN A 10.2.6.52
6638383363643661.headless.default.svc.cluster.local. 30	IN A 10.2.10.70
6337356433333633.headless.default.svc.cluster.local. 30	IN A 10.2.10.71

;; Query time: 2 msec
;; SERVER: 10.3.0.10#53(10.3.0.10)
;; WHEN: Wed Dec 14 20:48:41 UTC 2016
;; MSG SIZE  rcvd: 317
```